nibabel/filebasedimages.py

```diff
@@ -0,0 +1,569 @@
+    the datatype the header expects, setting any available header scaling
+    into the header to help the data match.
+
+    You can load the data into an image from file with::
+
+       img.from_filename(fname)
+
+    The image stores its associated files in its ``file_map`` attribute.  In order
+    to just save an image, for which you know there is an associated filename, or
+    other storage, you can do::
+
+       img.to_file_map()
+
+    You can get the data out again with::
+
+        img.get_data()
+
+    Less commonly, for some image types that support it, you might want to
+    fetch out the unscaled array via the object containing the data::
+
+        unscaled_data = img.dataoobj.get_unscaled()
+
+    Analyze-type images (including nifti) support this, but others may not
+    (MINC, for example).
+
+    Sometimes you might to avoid any loss of precision by making the
+    data type the same as the input::
+
+        hdr = img.header
+        hdr.set_data_dtype(data.dtype)
+        img.to_filename(fname)
+
+    Files interface
+    ===============
+
+    The image has an attribute ``file_map``.  This is a mapping, that has keys
+    corresponding to the file types that an image needs for storage.  For
+    example, the Analyze data format needs an ``image`` and a ``header``
+    file type for storage:
+
+       >>> import nibabel as nib
+       >>> data = np.arange(24, dtype='f4').reshape((2,3,4))
+       >>> img = nib.AnalyzeImage(data, np.eye(4))
+       >>> sorted(img.file_map)
+       ['header', 'image']
+
+    The values of ``file_map`` are not in fact files but objects with
+    attributes ``filename``, ``fileobj`` and ``pos``.
+
+    The reason for this interface, is that the contents of files has to
+    contain enough information so that an existing image instance can save
+    itself back to the files pointed to in ``file_map``.  When a file holder
+    holds active file-like objects, then these may be affected by the
+    initial file read; in this case, the contains file-like objects need to
+    carry the position at which a write (with ``to_files``) should place the
+    data.  The ``file_map`` contents should therefore be such, that this will
+    work:
+
+       >>> # write an image to files
+       >>> from io import BytesIO
+       >>> file_map = nib.AnalyzeImage.make_file_map()
+       >>> file_map['image'].fileobj = BytesIO()
+       >>> file_map['header'].fileobj = BytesIO()
+       >>> img = nib.AnalyzeImage(data, np.eye(4))
+       >>> img.file_map = file_map
+       >>> img.to_file_map()
+       >>> # read it back again from the written files
+       >>> img2 = nib.AnalyzeImage.from_file_map(file_map)
+       >>> np.all(img2.get_data() == data)
+       True
+       >>> # write, read it again
+       >>> img2.to_file_map()
+       >>> img3 = nib.AnalyzeImage.from_file_map(file_map)
+       >>> np.all(img3.get_data() == data)
+       True
+    '''
```

Comment:
Moved this to the `SpatialImage` class docstring. Also fixed up the docs; they were a bit hacked up.